src/i18n.js

```diff
@@ -0,0 +1,246 @@
+
+      let obj = travelDownObjectPath(this.registeredStrings[packageId], path);
+
+      if (isString) {
+        obj[last] = string;
+      } else if (!obj[last]) {
+        obj[last] = {};
+      }
+    });
+  }
+
+  getT(ns) {
+    if (!ns) return this.t;
+    return (key, formats) => this.t(`${ns}.${key}`, formats);
+  }
+
+  /**
+   * attempts to translate for a single language, given a preparsed path array.
+   * @return undefined if the language or string cannot be found,
+   *   and throws an error if the path isn't right.
+   */
+  tSingleLanguage(lang, _path, opts) {
+    let path = [..._path];
+
+    const ns = path.shift();
+    if (!ns) throw new Error(`key path seems invalid: [${_path.map(p => `"${p}"`).join(", ")}]`);
+
+    const languageObj = this.getLanguageObj(ns, lang);
+    if (!languageObj) return undefined;
+
+    const str = optionalTravelDownObjectPath(languageObj, path);
+    if (str) {
+      return this.format(ns, path, str, lang, opts);
+    } else {
+      return undefined;
+    }
+  }
+
+  /**
+   * gets a language object from a specified namespace
+   * @return undefined if it can't be found
+   */
+  getLanguageObj(ns, lang) {
+    return ns === "core"
+      ? this.getCoreLanguage(lang)
+      : this.getPkgLanguage(ns, lang);
+  }
+
+  /**
+   * gets a language for `core`
+   * @return undefined if it can't be found
+   */
+  getCoreLanguage(lang) {
+    const loaded = this.registeredStrings.core[lang]
+    if (loaded) return loaded;
+
+    const fetched = this.fetchCoreLanguageFile(lang);
+    if (!fetched) return undefined;
+
+    this.registeredStrings.core[lang] = fetched;
+    return fetched;
+  }
+
+  /**
+   * gets a language for a specific namespace
+   * @return undefined if it can't be found
+   */
+  getPkgLanguage(ns, lang) {
+    const loaded = this.registeredStrings[ns]?.[lang];
+    if (loaded) return loaded;
+
+    const fetched = this.fetchPkgLanguageFile(ns, lang);
+    if (!fetched) return fetched;
+
+    if (!this.registeredStrings[ns]) this.registeredStrings[ns] = {};
+    this.registeredStrings[ns][lang] = fetched;
+  }
+
+  /**
+   * fetches a core language from the disk
+   * @return undefined if it can't be found
+   */
+  fetchCoreLanguageFile(lang) {
+    let filepath = path.join(this.resourcePath, "i18n", `${lang}.json`);
+    let contents = JSON.parse(fs.readFileSync(filepath));
+
+    return contents;
+  }
+
+  /**
+   * fetches a language for a specific namespace
+   * @return undefined if it can't be found
+   */
+  fetchPkgLanguageFile(ns, lang) {
+    // TODO this could probably be optimised
+    let packages = this.packages.getAvailablePackages();
+    let foundPackage = packages.find(p => p.name === ns);
+
+    const i18nDir = path.join(foundPackage.path, "i18n");
+    const langfile = path.join(i18nDir, `${lang}.json`);
+
+    if (!(fs.isDirectorySync(i18nDir) && fs.existsSync(langfile))) return;
+
+    let contents = JSON.parse(fs.readFileSync(langfile));
```

Comment:
```suggestion
    const langfile = [ "json", "cson" ].map(ext -> `${lang}.${ext}`).find(f => fs.existsSync(path.join(i18nDir, f));
    
    if (!fs.isDirectorySync(i18nDir)) return;
    
    let contents;
    
    // Below we just grab the last index of the split to get the ext. 
    // Doing it this way allows any number of '.' to appear within a filename without failure
    switch(langfile.split(".")[langfile.split(".").length-1]) {
      case "json":
        contents = JSON.parse(fs.readFileSync(langfile));
        break;
      case "cson":
        CSON.readFile(langfile, (error, data) => {
          if (error) return;
          contents = data;
        });
        break;
    }
```

This is just to further point out what I meant in my above, if you add `const CSON = require("season");` at the top of this file, this code should work (but please feel free to actually check if you consider adding it) and allow packages to use JSON or CSON for their strings. At least to start, and this could always be expanded later on.

Reply:
mhm, makes sense!

Reply:
hm... I think I'm going to leave this one for now, and maybe implement it alongside [the ideas in the other comment](https://github.com/pulsar-edit/pulsar/pull/446#discussion_r1155498670), or something else? I feel like I want to have a think about this, and how we can elegantly make it seamless, but also extensible so users can use whatever arbitrary file format they want ~~(pon???)~~